libbpf-tools/funclatency.c

```diff
@@ -0,0 +1,333 @@
+
+	switch (key) {
+	case 'p':
+		errno = 0;
+		pid = strtol(arg, NULL, 10);
+		if (errno || pid <= 0) {
+			warn("Invalid PID: %s\n", arg);
+			argp_usage(state);
+		}
+		env->pid = pid;
+		break;
+	case 'm':
+		if (env->units != NSEC) {
+			warn("only set one of -m or -u\n");
+			argp_usage(state);
+		}
+		env->units = MSEC;
+		break;
+	case 'u':
+		if (env->units != NSEC) {
+			warn("only set one of -m or -u\n");
+			argp_usage(state);
+		}
+		env->units = USEC;
+		break;
+	case 'd':
+		errno = 0;
+		duration = strtol(arg, NULL, 10);
+		if (errno || duration <= 0) {
+			warn("Invalid duration: %s\n", arg);
+			argp_usage(state);
+		}
+		env->duration = duration;
+		break;
+	case 'i':
+		errno = 0;
+		interval = strtol(arg, NULL, 10);
+		if (errno || interval <= 0) {
+			warn("Invalid interval: %s\n", arg);
+			argp_usage(state);
+		}
+		env->interval = interval;
+		break;
+	case 'T':
+		env->timestamp = true;
+		break;
+	case 'h':
+		argp_state_help(state, stderr, ARGP_HELP_STD_HELP);
+		break;
+	case ARGP_KEY_ARG:
+		if (env->funcname) {
+			warn("Too many function names: %s\n", arg);
+			argp_usage(state);
+		}
+		env->funcname = arg;
+		break;
+	case ARGP_KEY_END:
+		if (!env->funcname) {
+			warn("Need a function to trace\n");
+			argp_usage(state);
+		}
+		if (env->duration) {
+			if (env->interval > env->duration)
+				env->interval = env->duration;
+			env->iterations = env->duration / env->interval;
+		}
+		break;
+	default:
+		return ARGP_ERR_UNKNOWN;
+	}
+	return 0;
+}
+
+static const char *unit_str(void)
+{
+	switch (env.units) {
+	case NSEC:
+		return "nsec";
+	case USEC:
+		return "usec";
+	case MSEC:
+		return "msec";
+	};
+
+	return "bad units";
+}
+
+static int attach_kprobes(struct funclatency_bpf *obj)
+{
+	struct bpf_link *l;
+	long err;
+
+	l = bpf_program__attach_kprobe(obj->progs.dummy_kprobe, false,
+				       env.funcname);
+	err = libbpf_get_error(l);
+	if (err) {
+		warn("failed to attach kprobe: %ld\n", err);
+		return -1;
+	}
+
+	l = bpf_program__attach_kprobe(obj->progs.dummy_kretprobe, true,
+				       env.funcname);
+	err = libbpf_get_error(l);
+	if (err) {
+		warn("failed to attach kretprobe: %ld\n", err);
+		return -1;
+	}
+
+	return 0;
+}
+
+static int attach_uprobes(struct funclatency_bpf *obj)
+{
+	char *binary, *function;
+	char bin_path[PATH_MAX];
+	off_t func_off;
+	int ret = -1;
+	struct bpf_link *l;
+	long err;
+
+	binary = strdup(env.funcname);
+	if (!binary) {
+		warn("strdup failed");
+		return -1;
+	}
+	function = strchr(binary, ':');
+	if (!function) {
+		warn("Binary should have contained ':' (internal bug!)\n");
+		return -1;
+	}
+	*function = '\0';
+	function++;
+
+	if (resolve_binary_path(binary, env.pid, bin_path, sizeof(bin_path)))
+		goto out_binary;
+
+	func_off = get_elf_func_offset(bin_path, function);
+	if (func_off < 0) {
+		warn("Could not find %s in %s\n", function, bin_path);
+		goto out_binary;
+	}
+
+	l = bpf_program__attach_uprobe(obj->progs.dummy_kprobe, false,
```

Comment:
ok, so somehow I didn't pay attention to this initially.

You are leaking link here (and you are not explicitly detaching it, neither on error, nor in the normal flow).

You can store returned link into the skeleton like this:

```
obj->links.dummy_kprobe = bpf_program__attach_uprobe(...);
```

Later, when you destory the skeleton, those links will be destroyed as well. It's fine if link themselves failed (so no need to NULL them out)

Reply:
thanks.  i was just using 'l' for the error check; didn't know they needed to be set like that.

fixed and pushed.